### DISCUSSION TITLE
Proposal: Midi Effects Suite

### DISCUSSION THREAD

**Discussion by legsmechanical** [SynthstromAudible/DelugeFirmware#466] [Ideas]
This proposes adding a suite of midi effects that can creatively and musically manipulate midi data in real time (mostly). 

This proposal is based heavily on the "PLAY FX" functions of the old Yamaha [RS7000](https://usa.yamaha.com/files/download/other_assets/0/319360/RS7000E1.pdf) and RM1x, which was creative dynamite that I've yet to see replicated outside of Squarp sequencers.

I'm not a dev, but hope this can inspire any particularly ambitious ones. :)

-------

**General functional notes**:
- These effects operate on a per-clip level (though perhaps there's some utility for an "affect all" implementation? It could get messy and start racking up the voice count very quickly.)
- Effects operate on recorded sequencer data and/or live-input midi (either from the deluge's audition pads/keyboard mode/or external midi), as specified in the effect block details below.
- Each clip has its own set of the 4 effect blocks described below. The parameters of the 4 blocks are retained and unique to each clip. 
- The 4 effect "blocks" are  routed serially and feed into each other.  Ideally, the order in which the blocks process and feed data into subsequent blocks is user-definable - but if they're static, some thought will need to go into the most useful configuration.
- Effects manipulate incoming sequencer data in real-time and are non-destructive.  I.e., when they're turned "off" the underlying sequencer data plays at it was originally programmed/recorded.
- **However**, when overdubbing (REC>PLAY) while any combination of the effects is turned on, the underlying clip data is replaced, as it plays, with the midi data generated by the final effect block in the chain (including any pass-through notes - see below).  So you can manipulate the data non-destructively in realtime but then "print" the data destructively to the clip if you like (kind of a midi equivalent to resampling).  (Probably some challenges here with respect to clip length and the STRETCH parameter in the proposed OFFSET block.) 
- Once the data is “printed” via overdub, each effect block is reset to the “off” position and the effect parameters are reset to their default values. 
- Effect parameters are learnable to the gold knobs (except for some obvious exceptions where that wouldn't be particularly useful).
- Effect parameters that manipulate timing respect the current swing setting (which is itself a sort of midi effect). 
- The RS7k's manual (linked above) section on "Play FX" might clarify some of the parameter functionality, particularly with respect to scaling in the proposed OFFSET block.

**General UI notes:**

- The effects are laid out in 4 unique "blocks" per clip.  For the most part, each "block" changes various parameters for a single new "effect."  The exception is the "Offset" block, which houses a number of effects that require fewer parameters. 
- The UI proposed below is based on that in the [community master compressor menu](https://youtu.be/zzeIdRB9PK8?si=2cCIwHvqKGFhwVVj&t=70) and would function in a similar way.  Up to 6 parameters viewable on the OLED at the same time per effect "block."  (This is reflected in the diagrams accompanying each block description below).
- Entering the Effects blocks: I'd propose something like holding SYNC-SCALING while in clip view and pressing the SELECT knob to view the first block and then holding SYNC-SCALING and turning SELECT to switch between blocks.
-  For parameter editing, I'd propose pushing SELECT to switch between visible parameters and turning SELECT to adjust them (all while already in effects block-view).

-------

**Overview of effect blocks**
The following describes the proposed operation of each effect block's parameters, with diagrams approximating how the block would be displayed on OLED.

**CHORD/STRUM**
Adds up to four unison or "strummed" notes to the incoming to generate chords & strums

Off/Unison/(Strum sync)| --
-- | --
Note 1 | Note 2
Note 3 | Note 4

-  OFF/CHORD/(STRUM SYNC) changes current mode.  In UNISON mode, all generated notes (defined by NOTE1 thru NOTE 4) are played at the same time.  After scrolling past UNISON, various "strum sync" values are displayed (1/64; 1/32; 1/16; 1/8; 1/4, along with their dotted/triplet values).  When a strum sync value is selected, the playback of each of the 4 generated notes  is sequentially delayed by the selected time division.
- NOTE value is +/- 24 steps with "Orig" and "Mute" at the neutral position.  The NOTE value refers to the intervals between notes within the clip's currently selected scale.  (I.e., the generated notes are always constrained to the clip's scale/key.)   E.g., with NOTE1 at 2, NOTE2, at +7, NOTE3 at -7, and NOTE4 at +5, the additional notes generated by a C3 when the selected clip scale is Cmaj would be E3, C4, C2, & A3. 
- If a NOTE value is set to "Orig," the original input note is played again at the appropriate time division when a strum sync value is selected.  "Orig" has no effect in CHORD mode.
- If a NOTE value is set to the "Mute" position, it will not be played but will be taken in account for timing purposes when a "strum sync" value is selected.  E.g, with a  strum sync value of 1/8, and NOTE1 at +7, NOTE2 at -7, NOTE3 at "off" and NOTE4 at +7, a C3 note would would generate a C4 1/8 beat after the C3, a C2 another 1/8 beat later, an 1/8 beat rest another 1/8 beat later, and then a C4.  
- This effect operates on both sequencer data and live midi input while the sequencer is running.
- The original underlying note(s) are "passed through" for the purpose of feeding other blocks and printing the data in overdub.

**MIDI DELAY**
Generates additional sequentially delayed notes based on the incoming note value.

OFF/ON |  --
-- | --
Feedback | Interval
Sync | Velocity

- FEEDBACK selects the number of delayed notes to generate (would probably want to limit this to around 8 or so, but who knows?)
- SYNC indicates the sequential delay of each generated note.  Values: 1/64; 1/32; 1/8; 1/4, along with their dotted/triplet values.
- INTERVAL (value +/- 12 steps with OFF in the neutral position) changes the sequential note interval of generated notes.  If the value is positive, the interval of each generated note is sequentially increased by the value; if negative, the intervals are sequentially decreased.  The INTERVAL value refers to the intervals between notes within the clip's currently selected scale.   (I.e., the generated notes are always constrained to the clip's scale/key.)  E.g.,  With a FEEDBACK value of 3, an INTERVAL value of +2, and the clip scale set to Cmaj, the additional delayed notes generated by a C2 would be E2, G2, B2.  
- VELOCITY (value +/- 50, OFF in the neutral position) sequentially scales up/down the velocity of generated notes to create fade-in/fade-out effects.   
- This effect operates on both sequencer data and live midi input while the sequencer is running.
- The original underlying note(s) are "passed through" for the purpose of feeding other blocks and printing the data in overdub.

**ARP**
 Arpeggiates incoming notes

On/off| --
-- | --
Mode | Octave
Sync| Gate

- MODE selects standard arp modes (up; down; up/down; down/up; etc. - note: would be cool to see other arp algorithms like thump-down, pinky-up, etc.
- OCTAVE does what arp octaves do
- SYNC selects arp timing (usual time divisions)
- GATE selects the length of generated notes.
- NOTE this is not meant to replace the OG arpeggiator function, but to supplement it with a version that can "print" the generated notes to the clip and operate pre/post other effect blocks.
- This effect operates on both sequencer data and live midi input while the sequencer is running.
- Only incoming notes that do not overlap with other notes are passed through.

**OFFSETS**
Makes basic scaled/absolute changes to incoming midi information

Transpose| Stretch
-- | --
Velocity | Absolute (velocity)
Gate | Absolute (gate)

- TRANSPOSE (value +/- 24 steps, with OFF in the neutral position) transposes incoming notes up or down by the selected number of steps.  The value refers to the intervals between notes within the clip's currently selected scale.   (I.e., the generated notes are always constrained to the clip's scale/key.)  TRANSPOSE operates only on sequenced data.  Incoming notes are not passed through.
- STRETCH (value: /2, /4, /8, OFF, *2, *4, *8) compresses or expands the timing of incoming notes by scaling the timing between them up or down relative to the selected division/multiplication value. It is essentially a real-time per-clip tempo multiplier/divider. [some challenges here to think through, I assume].  STRETCH operates only on sequenced data.  Incoming notes are not passed through.
- VELOCITY (value: +/- 50 with "off" in the neutral position) scales the velocity of incoming notes up or down relative to each other.  GATE operates only on sequenced data, except when ABSOLUTE is "on."
- ABSOLUTE (velocity) (On/Off) changes how velocity behaves.  With ABSOLUTE "off,"  velocity is scaled as described above.  With ABSOLUTE "on," the VELOCITY values change to 0-127, and set the absolute velocity of all incoming notes to the selected value.
-  GATE (value: +/- 50, with off in the neutral position) scales the length (time to note-off) of incoming notes up or down relative to one another.  GATE operates only on sequenced data, except when ABSOLUTE is "on" and the sequencer is running.
-  ABSOLUTE (gate) changes how GATE behaves.  With ABSOLUTE "off,"  note length is scaled as described above.  With ABSOLUTE "on," the GATE values change from 1/128 to 1 bar, and set the absolute note length of all incoming notes to the selected value.

**Comment by dailyherold** [comment]
I like the ideas! Been watching RetroKits videos of their RK-008, RK-006, and RK-002 devices. Some more MIDI wizardry ideas there. Wondering if [exploding](https://www.youtube.com/watch?v=_1D_TEMdg5I) and [merging](https://www.youtube.com/watch?v=64CAOUHLnjM) would be handy for the Deluge? Seems very helpful for the RK-008's 8-track constraint, but maybe helpful regardless for us, e.g. turning one drum/perc midi clip into multiple clips based on notes played. Merging multiple midi clips from same device as well.

Note: I'm a new Deluge user, so might be more efficient ways of doing this beyond what I'd guess is the workflow alternative to "exploding": clip duplicate per unique note and removing _other_ notes per clip. Not sure how I'd "merge" today though, found [this related forum post](https://forums.synthstrom.com/discussion/3824/merging-consolidating-sequencer-clips) with similar conundrum.

**Comment by legsmechanical** [comment]
(And no reason not to add a quantize/humanize block as well… )

**Comment by runawaykiwi** [comment]
Demonstration of non-destructive chained MIDI effects on Hapax: https://youtu.be/tTLQx2kexNM?t=1564

Demonstration of destructive midi effects on Hapax: https://youtu.be/tTLQx2kexNM?t=1696

**Comment by thejesiah** [comment]
Correct me if I'm wrong, but the Korg Volca Beats, for as basic as it is, the way it integrates the Stutter function on the MIDI side of things is completely unique at least for hardware in that pricerange. It's really fun and something I miss on much more expensive gear... seems like it's kind of one of the FX that could be implemented with what OP is proposing. 

Likewise, the OP-Z has Step Components that manipulate the sequencer/MIDI data in a way that isn't reproducible downstream. 

Whatever the community devs implement, it'll be huge for sound/track design on the Deluge.

**Comment by gramster** [comment]
I think if the experienced devs could implement the core functionality here with a couple of these effects, this could open up to other contributors if they had a couple of examples of MIDI effects to work from. Also no need to just have the four above; the Hapax has a whole bunch documented here https://squarp.net/hapax/manual/modefx/. Ableton Live has quite a few too; may be some sources of inspiration there.

Would really love to see this. That said, its an area where the OLED screen would really be useful and the 7seg would be a challenge.
